evm_loader/program/src/evm/database.rs

```diff
@@ -51,3 +49,44 @@ pub trait Database {
         is_static: bool,
     ) -> Option<Result<Vec<u8>>>;
 }
+
+/// Provides convenience methods that can be implemented in terms of `Database`.
+#[maybe_async(?Send)]
+pub trait DatabaseExt {
+    /// Returns whether an account exists, which is also referred to as the account being non-empty,
+    /// in accordance with https://eips.ethereum.org/EIPS/eip-161.
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool>;
+
+    /// Returns the code has for an address in accordance with https://eips.ethereum.org/EIPS/eip-1052.
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]>;
+}
+
+#[maybe_async(?Send)]
+impl<T: Database> DatabaseExt for T {
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool> {
+        Ok(self.nonce(address, chain_id).await? > 0 || self.balance(address, chain_id).await? > 0)
+    }
+
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]> {
+        // `self.code` will return an empty buffer when the account does not exist, and when the
+        // account exists but does not have any code. For that reason we also have to check if the
+        // account exists if the buffer is empty since we must return [0u8; 32] if it does not. We
+        // could check if the account exists first, but that would lead to more computations in the
+        // common case where the account exists and contains code.
```

Comment:
While I agree we should have more comments in some places of the code base, I believe this is too verbose. I suggest removing all comments from this trait and its implementation, apart from the links to the EIPs.

Reply:
I think comments are useful to describe things that are not immediately obvious from the code. So the comment here that talks about why the implementation is written a certain way I think is useful. It can probably be written better though. 

The other comments are not so useful, they can be easily deducted from the implementation. Especially for "account exists" where we abstracted the computation with a function that has a decent name.